rfcs/2021-08-13-8025-internal-tracing.md

```diff
@@ -0,0 +1,254 @@
+between high-level metrics and low-level CPU profiling.
+
+### What is the impact of not doing this?
+
+Performance investigations would continue to rely on a combination of intuition
+and expert-level tooling to narrow down problems reported by users. Instead of
+having data directly indicating that, for example, a component is spending all
+of its time fingerprinting files, we would need to deduce that from other
+sources of information.
+
+### How does this position us for success in the future?
+
+While this is far from an endgame tracing system, it puts a number of important
+components into place and helps us start gaining experience with them. Over
+time, as we fix issues and gain confidence in this relatively simple setup, we
+can add additional layers of more detailed spans, measure and optimize overhead,
+experiment with options like sampling, etc. If all goes well, we could get to
+a significantly more advanced system with small iterative steps from this base.
+
+## Drawbacks
+
+### Why should we not do this?
+
+An argument could be made that the combination of coupling our logging and
+tracing, and the relative immaturity of the community crates for shipping traces
+to Datadog will result in a somewhat fragile system. This has shown itself to
+a degree with the memory growth issues in early testing, and it's not entirely
+clear how we'd go about observing the trace forwarding libraries themselves. The
+counter argument would be that this use of off-the-shelf components is exactly
+what makes this initial experimentation cheap, we have a plan to avoid the
+issues we've run into, and we'll have plenty of time to flush out issues in our
+own soak test environments before recommending this to users.
+
+Another, somewhat more vague objection could involve the runtime characteristics
+of Vector. Tracing largely comes from request-response oriented systems, where
+there is a very clear unit of work around which to wrap your tracing spans. As
+a streaming system, Vector does not have as clear a boundary for those spans.
+This mismatch has shown itself as part of the memory growth issue with
+long-lived spans. It's possible that some components may not have a clear
+concept of a single "iteration" around which we can wrap a base span, and we'll
+have a hard time instrumenting them in a way that is both logical and
+practically useful.
+
+### What kind on ongoing burden does this place on the team?
+
+The ongoing maintenance burden should be quite minimal. The extent of code
+changes would be the addition of some small span instantiations and
+instrumentation, but nothing that would require threading through existing
+interfaces or reorganizing any abstractions.
+
+## Prior Art
+
+There is very little prior art available for tracing streaming system, but we
+are largely just using existing off-the-shelf components. We've not opted to
+invent anything particularly new for Vector.
```

Comment:
One thing that had come up in my profiling reading struck a chord here: user-space dtrace probes. Downsides with these are there's no support for some of our deploy targets -- Windows -- but we avoid any runtime impact that tracing currently imparts. Oxide is using them in their work: https://github.com/oxidecomputer/usdt